lib/frontend/src/switch.rs

```diff
@@ -0,0 +1,193 @@
+use std::collections::HashMap;
+use cranelift_codegen::ir::*;
+use cranelift_codegen::ir::condcodes::IntCC;
+use frontend::FunctionBuilder;
+
+type EntryIndex = i64;
+
+#[derive(Debug)]
+pub struct Switch {
+    cases: HashMap<EntryIndex, Ebb>
+}
+
+impl Switch {
+    pub fn new() -> Self {
+        Switch {
+            cases: HashMap::new(),
+        }
+    }
+
+    pub fn set_entry(&mut self, index: EntryIndex, ebb: Ebb) {
+        assert!(self.cases.insert(index, ebb).is_none(), "Tried to set the same entry {} twice", index);
+    }
+
+    fn build_cases_tree(self) -> Vec<(EntryIndex, Vec<Ebb>)> {
+        println!("before: {:#?}", self.cases);
+        let mut cases = self.cases.into_iter().collect::<Vec<(_, _)>>();
+        cases.sort_by_key(|&(index, _)| index);
+
+        let mut cases_tree: Vec<Vec<(EntryIndex, Ebb)>> = vec![];
+        let mut last_index = None;
+        for (index, ebb) in cases {
+            if Some(index - 1) > last_index {
+                cases_tree.push(Vec::new());
```

Comment:
Would it work to make `cases_tree` be a `Vec<(EntryIndex, Vec<ebb>)>`, and do `cases_tree.push((index, Vec::new())` here? If so, it seems like you could avoid the `map` step below.